tests/php/RedirectFileControllerTest.php

```diff
@@ -0,0 +1,453 @@
+            false,
+            $response,
+            'Old Hash URL of unpublsihed files should return 404'
+        );
+    }
+
+    public function testRedirectAfterDeleting()
+    {
+        $file = File::find('FileTest-subfolder/FileTestSubfolder.txt');
+        $file->publishSingle();
+        $v1Url = $file->getURL(false);
+
+        $file->setFromString('version 2', $file->getFilename());
+        $file->write();
+        $file->publishSingle();
+        $v2Url = $file->getURL(false);
+
+        $file->delete();
+        $file->deleteFile();
+
+        $response = $this->get($v2Url);
+        $this->assertResponse(
+            404,
+            '',
+            false,
+            $response,
+            'Deleted file file should return 404'
+        );
+
+        $response = $this->get('/assets/FileTest-subfolder/FileTestSubfolder.txt');
+        $this->assertResponse(
+            404,
+            '',
+            false,
+            $response,
+            'Legacy URL of deleted files should return 404'
+        );
+
+        $response = $this->get($v1Url);
+        $this->assertResponse(
+            404,
+            '',
+            false,
+            $response,
+            'Old Hash URL of deleted files should return 404'
+        );
+    }
+
+    public function testVariantRedirect()
+    {
+        /** @var Image $file */
+        $file = Image::create();
+
+        $file->FileFilename = 'test.jpg';
+        $file->setFromLocalFile(__DIR__ . '/ImageTest/landscape-to-portrait.jpg', 'test.jpg');
+        $file->write();
+        $file->publishSingle();
+        $ico = $file->ScaleWidth(32);
+        $icoUrl = $ico->getURL(false);
+        $suffix = $ico->getVariant();
+
+        $response = $this->get($icoUrl);
+        $this->assertResponse(
+            200,
+            $ico->getString(),
+            false,
+            $response,
+            'Publish variant sghould resolve with 200'
+        );
+
+        $response = $this->get("/assets/test__$suffix.jpg");
+        $this->assertResponse(
+            307,
+            '',
+            $icoUrl,
+            $response,
+            'Legacy path to variant should redirect.'
+        );
+
+        $response = $this->get("/assets/_resampled/$suffix/test.jpg");
+        $this->assertResponse(
+            307,
+            '',
+            $icoUrl,
+            $response,
+            'SS3 Legacy path to variant should redirect.'
+        );
+
+        $file->setFromLocalFile(__DIR__ . '/ImageTest/test-image-high-quality.jpg', 'test.jpg');
+        $file->write();
+        $file->publishSingle();
+        $ico = $file->ScaleWidth(32);
+        $icoV2Url = $ico->getURL(false);
+
+        $response = $this->get($icoUrl);
+        $this->assertResponse(
+            307,
+            '',
+            $icoV2Url,
+            $response,
+            'Old URL to variant should redirect with 307'
+        );
+    }
+
+    public function testVariantInFolderRedirect()
```

Comment:
Same here, please test with multiple folder depth and underscores in file/folder name

Reply:
The tests in this method are now covered by the data provider in `testVariantRedirect`. So I removed the method.